### PR TITLE
WebGLRenderer: Integrate PMREM.

### PR DESCRIPTION
**Pull request by Mugen87**
Related issue: see #20463

**Description**

Second attempt of integration PMREM.

### CODE REVIEW

**Comment by WestLangley** [comment]
Your auto-generated PMREM implementation is working for me. Thanks for the PR!

Personally, I would focus only on auto-generated PMREM environment maps for now and save the background blurriness for another PR.

PMREM is too blocky as a blurry background, and not even close to linear in your "alpha" parameter when used that way.

@elalish may have some suggestions for improvement -- and recommendations on how to handle the "roughness-mipmapper" feature.

**Comment by elalish** [comment]
> @elalish may have some suggestions for improvement -- and recommendations on how to handle the "roughness-mipmapper" feature.
The biggest improvement to make is supporting variable-size PMREMs, which I have been meaning to get around to for ages. That could also help with the blockiness problem. The RoughnessMipmapper is pretty independent of this PR, but agree that building the functionality into the renderer would be great. It helps a lot with aliasing and general PBR correctness when zooming out.

**Comment by elalish** [comment]
I'm not really sure how to even define linearity for blurriness, especially on a 0-1 scale. If you do allow variable-size PMREMs eventually, you'll find that to get 1-mip down from full, you'll need increasingly tiny values or roughness (blurriness). But on the other hand, at a given value you'll always get the same output regardless of input resolution. So it depends how you want it to behave.

**Comment by WestLangley** [comment]
>The biggest improvement to make is supporting variable-size PMREMs, which I have been meaning to get around to for ages

Hmm... I have not seen a need for that. I suppose you are aware of a need?

//

It is personal taste, obviously, but a non-blocky, "slightly out-of-focus" blurry would be welcome -- like a depth-of-field pass.

As it is currently implemented, I think PMREM should only be used for what it was designed for.

**Comment by elalish** [comment]
I was thinking it could help with performance to reduce the PMREM texture size for models that don't have low roughness. On the other side, higher res would make for better backgrounds. Obviously it hasn't been a huge priority...

**Comment by WestLangley** [comment]
>I was thinking it could help with performance to reduce the PMREM texture size for models that don't have low roughness.

That is what I was guessing: you were referring to _reducing_ the size and the levels.

Has there been a performance problem? If so, it should be noted.

**Comment by elalish** [comment]
> Has there been a performance problem? If so, it should be noted.

Well, we always want to increase performance, right? What I have not measured yet is what gain might be available from a change like this. If you have any performance numbers that might shed some light, they would be much appreciated!

**Comment by Mugen87** [comment]
> Personally, I would focus only on auto-generated PMREM environment maps for now and save the background blurriness for another PR.

Then we need to find a way to solve the issue I have mentioned in https://github.com/mrdoob/three.js/pull/20463#issuecomment-731727229.

**Comment by elalish** [comment]
@Mugen87 
> Meaning if the feature is missing, the background will lock blocky because of the nearest filtered HDR textures.

This is an interesting comment. Is this the current problem? Because as long as you use the cubeuv shader fragment, you should still be able to get interpolated values from the PMREM. Maybe we just need to add that to Background?

**Comment by WestLangley** [comment]
@Mugen87 May be missing something... I am not sure why PMREM is used for backgrounds in the first place. It should't be in my opinion.

If the HDR is blocky, then you need to avoid `THREE.UnsignedByteType` so linear filtering can be used.

**Comment by Mugen87** [comment]
I've seen many users assigning the result of `PMREMGenerator` to `Scene.environment` and `Scene.background`. We also do this in more than one example right now.

If the renderer starts to provide an auto-conversion, `WebGLBackground` needs access to the the result of `PMREMGenerator`, too. Otherwise users have to work with `PMREMGenerator` on application level again to achieve the same background as before. The goal is however that user should not be confronted with `PMREMGenerator` at all. So this is a backwards compatibility issue.

**Comment by WestLangley** [comment]
>I've seen many users assigning the result of PMREMGenerator to Scene.environment and Scene.background. We also do this in more than one example right now.

Well, that is not correct. PMREM is an IBL, not a background image.

**Comment by Mugen87** [comment]
Just to be clear I'm referring on code like:

https://github.com/mrdoob/three.js/blob/3a9152af37e7879e5ed41cb60c549e6a78df57aa/examples/webgl_tonemapping.html#L103-L106

**Comment by elalish** [comment]
> If the HDR is blocky, then you need to avoid `THREE.UnsignedByteType` so linear filtering can be used.

Not true! The way CubeUV works for PMREM is that it does in fact use nearest filtering, but the `textureCubeUV` shader function does manual trilinear interpolation (since we don't have hardware interpolation support for RGBE). So if the background looks like nearest, then there's a bug.

**Comment by WestLangley** [comment]
>If the HDR is blocky, then you need to avoid THREE.UnsignedByteType so linear filtering can be used.
>>Not true! The way CubeUV works for PMREM is ...

I wasn't referring to PMREM.

**Comment by Mugen87** [comment]
So the be clear the example code should actually look like so, right? https://jsfiddle.net/pg5rhctx/2/

**Comment by WestLangley** [comment]
>So the be clear the example code should actually look like so, right? https://jsfiddle.net/pg5rhctx/2/

Yes, IMO, the rendering should look like the rendering in your fiddle.

The app in your fiddle should not have to handle PMREM.

```js
scene.background = texture; // the loaded cubeMap or equirectangular texture

scene.environment = texture; // environment texture will be auto-converted to PMREM format by the renderer

new GLTFLoader()
	.setPath( '...' )
	.load( 'DamagedHelmet.gltf', function ( gltf ) {

		scene.add( gltf.scene ); // the app does not have to do anything else (yay!)
		render();

	} );

} );
```

No more PMREM-related code at the app level.

**Comment by Mugen87** [comment]
> The app in your fiddle should not have to handle PMREM.

Yes, I just wanted to make sure we clarify how `Scene.background` is configured 👍 .

**Comment by Mugen87** [comment]
I've updated most examples like suggested in https://github.com/mrdoob/three.js/pull/22178#issuecomment-886058226. The ones which are using `RoomEnvironment` still have to access `PMREMGenerator` in order to generate the environment map.

For a different PR, we might want to consider to allow `Scene` as a type for `Scene.environmet`. In this way, the renderer could internally use `PMREMGenerator.fromScene()` and this code would disappear from app level.

File: examples/webgl_loader_gltf.html [hunk 1]
```diff
@@ -40,17 +40,14 @@
 				scene = new THREE.Scene();
 
 				new RGBELoader()
-					.setDataType( THREE.UnsignedByteType )
+					.setDataType( THREE.FloatType )
```
Comment (Mugen87) [inline examples/webgl_loader_gltf.html:43]: Should we consider to make `THREE.FloatType` the default similar to `EXRLoader`? I guess most users do not want the nearest filtered background.

Comment (WestLangley) [reply examples/webgl_loader_gltf.html:43]: Short-answer: yes.

`UnsignedByteType` was made the default so as to support mobile devices under WebGL1.

It is now several years later, and it appears that linear-float filtering is well-supported in WebGL1, and by default in WebGL2.

I was initially concerned about the change to `THREE.FloatType` in 12 examples in this PR. But if the support is now widely-available, then that is OK, too.

Your call.

Comment (Mugen87) [reply examples/webgl_loader_gltf.html:43]: Let's make a separate PR for this change.

**Comment by WestLangley** [comment]
@mrdoob This looks good to me.

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
@Mugen87 

> For a different PR, we might want to consider to allow `Scene` as a type for `Scene.environmet`. In this way, the renderer could internally use `PMREMGenerator.fromScene()` and this code would disappear from app level.

Interesting idea... The serialisation code would have to be redesigned a bit in order to support that though...

**Comment by WestLangley** [comment]
On LEFT:
```js
const pmremGenerator = new THREE.PMREMGenerator( renderer );
const hdrCubeRenderTarget = pmremGenerator.fromEquirectangular( equirectHDR );
material.envMap = hdrCubeRenderTarget.texture; // PMREM texture
```
On RIGHT:
```js
material.envMap = equirectHDR; // loaded texture
```
Before
Screen Shot 2021-07-27 at 9 33 33 PM

After
Screen Shot 2021-07-27 at 9 33 53 PM

For users who were not using PMREM with `MeshStandardMaterial` or `MeshPhysicalMaterial`, illumination will now be brighter. Users may now find they can reduce -- or remove -- scene-added `AmbientLight`.

**Comment by mrdoob** [comment]
@WestLangley 

Good catch, if someone was was just only using `envMap` now they'll have IBL on top.

I guess they'll have to switch to `MeshMatcapMaterial`?

**Comment by WestLangley** [comment]
@mrdoob I do not understand you comment, TBH.

`MeshStandard/PhysicalMaterial` is now computing proper illumination even without the user assigning a PMREM.

**Comment by mrdoob** [comment]
> @mrdoob I do not understand your comment, TBH.

Indeed. Switching to `MeshMatcapMaterial` doesn't make sense... 😅

> `MeshStandard/PhysicalMaterial` is now computing proper illumination even without the user assigning a PMREM.

Well... I guess we'll have to wait and see if anyone complains.

**Comment by WestLangley** [comment]
Maybe you are misunderstanding what has happened... (?) There are no complaints.

Assigning an ordinary cube map as `envMap` previously produced incorrect results because it did not account for the irradiance implied by the `envMap`. Previously, the user had to either create a PMREM map and use that as the `envMap` -- or add `AmbientLight` to the scene to compensate.

Now, users should not use PMREM at all. And they do not need to. We can remove PMREM from the examples.

Their scenes will be brighter that before. Tell them to remove the unneeded ambient lighting.

**Comment by makc** [comment]
> > The biggest improvement to make is supporting variable-size PMREMs, which I have been meaning to get around to for ages
> 
> Hmm... I have not seen a need for that. I suppose you are aware of a need?
> 
 
I was going to quote https://github.com/google/model-viewer/issues/912 but apparently after almost a year someone had fixed it

**Comment by mrdoob** [comment]
> @mrdoob Maybe you are misunderstanding what has happened... (?) There are no complaints.

Ah, I misunderstood indeed. Yeah... I didn't think of that scenario.
Good to be aware in case someone mentions it 👍

**Pull request merged**